pkg/protocol/engine/congestioncontrol/scheduler/drr/scheduler.go

```diff
@@ -0,0 +1,446 @@
+			// TODO: provide options properly rather than defaults here
+			optsRate:                           100,
+			optsMaxBufferSize:                  100,
+			optsAcceptedBlockScheduleThreshold: 10 * time.Second,
+			optsMaxDeficit:                     100,
+			optsMinMana:                        1,
+			optsTokenBucketSize:                1000,
+		}, opts,
+	)
+}
+
+func (s *Scheduler) Shutdown() {
+	close(s.shutdownSignal)
+	s.TriggerStopped()
+}
+
+// Start starts the scheduler.
+func (s *Scheduler) Start() {
+	s.shutdownSignal = make(chan struct{}, 1)
+	s.blockChan = make(chan *blocks.Block, 1)
+	go s.mainLoop()
+
+	s.TriggerInitialized()
+}
+
+// Rate gets the rate of the scheduler in units of work per second.
+func (s *Scheduler) Rate() int {
+	return s.optsRate
+}
+
+func (s *Scheduler) IsBlockIssuerReady(accountID iotago.AccountID, blocks ...*blocks.Block) bool {
+	// if the buffer is completely empty, any issuer can issue a block.
+	if s.buffer.Size() == 0 {
+		return true
+	}
+	work := 0
+	// if no specific block(s) is provided, assume max block size
+	// TODO: Define max block work instead of using size
+	if len(blocks) == 0 {
+		work = iotago.MaxBlockSize
+	}
+	for _, block := range blocks {
+		work += block.Work()
+	}
+	deficit, err := s.getDeficit(accountID)
+	if err != nil {
+		return false
+	}
+
+	return deficit >= uint64(work+s.buffer.IssuerQueue(accountID).Work())
+}
+
+func (s *Scheduler) getDeficit(accountID iotago.AccountID) (uint64, error) {
+	s.deficitsMutex.RLock()
+	defer s.deficitsMutex.RUnlock()
+
+	d, exists := s.deficits.Get(accountID)
+	if !exists {
+		_, err := s.manaRetrieveFunc(accountID)
+		if err != nil {
+			return 0, ierrors.Errorf("%w: no deficit or Mana available for issuer %s", err, accountID)
+		}
+		// load with max deficit if the issuer has Mana but has been removed from the deficits map
+		return s.optsMaxDeficit, nil
+	}
+
+	return d, nil
+}
+
+func (s *Scheduler) AddBlock(block *blocks.Block) {
+	s.bufferMutex.Lock()
+	defer s.bufferMutex.Unlock()
+
+	droppedBlocks, err := s.buffer.Submit(block, s.manaRetrieveFunc)
+	// error submitting indicates that the block was already submitted so we do nothing else.
+	if err != nil {
+		return
+	}
+	for _, b := range droppedBlocks {
+		b.SetDropped()
+		s.events.BlockDropped.Trigger(block)
+	}
+	block.SetEnqueued()
+	s.tryReady(block)
+}
+
+func (s *Scheduler) mainLoop() {
+	var blockToSchedule *blocks.Block
+loop:
+	for {
+		select {
+		// on close, exit the loop
+		case <-s.shutdownSignal:
+			break loop
+		// when a block is pushed by the buffer
+		case blockToSchedule = <-s.blockChan:
+			tokensRequired := float64(blockToSchedule.Work()) - (s.tokenBucket + float64(s.optsRate)*time.Since(s.lastScheduleTime).Seconds())
+			if tokensRequired > 0 {
+				// wait until sufficient tokens in token bucket
+				timer := time.NewTimer(time.Duration(tokensRequired/float64(s.optsRate)) * time.Second)
+				<-timer.C
+			}
+			s.tokenBucket = lo.Min(
```

Comment:
How does resetting tokenBucket work? It seems like at first it will grow, but after it exceeds `optsTokenBucketSize` value, it can never go below that (1000 by default). Is that intentional?

Reply:
I forgot a crucial line in the token bucket update to deduct the work for each scheduled block. This is now included when a block is scheduled.